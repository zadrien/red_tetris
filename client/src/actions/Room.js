export const ROOM = "ROOM"

export const onJoined = () => ({
  event: "JOINED",
  handle: ROOM,
  state: "JOINED"
})

export const onDisplay = () => {
    return {
	event: "DISPLAY",
	handle: ROOM,
	state: "DISPLAY",
    }

}

export const onHost = () => {
    return {
	event: "HOST",
	handle: ROOM,
	state: "HOST",
    }
}

export const onPlayers = () => {
    return {
	event: "PLAYERS",
	handle: ROOM,
	state: "PLAYERS",
    }
}

export const onStart = () => ({
  event: "START",
  handle: ROOM,
  state: "START"
})

export const emitStart = (id) => {
  return {
    event: "START",
    emit: true,
    payload: id
  }
}


export const emitMove = (move) => ({
    event: "CONTROLLER",
    emit: true,
    payload: move
})


export const startGame = (start) => {
  return {
    type: "ROOM",
    state: "START",
    start
  }
}

export const onGameOver = () => ({
	event: "GAMEOVER",
	handle: "ROOM",
	state: "GAMEOVER"
})

export const emitQuit = (room) => ({
  emit: true,
  event: "LEAVE",
  payload: room
})

export const onQuit = () => ({
  event: "LEAVE",
  handle: ROOM,
  state: "QUIT",
})

export const removeOverlay = () => ({
  type: ROOM,
  state: "RESET",
  reset: true
})
