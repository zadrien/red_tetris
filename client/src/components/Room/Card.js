import React from 'react'
import { connect } from 'react-redux'

const style = {
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  
  width: "auto",
  height: "236px",
  backgroundColor: "#dd4545",
  padding: "5px",
  borderRadius: "3px",

  marginRight: "5px"
}

const color = {
  ".": "Black",
  C: "Cyan",
  B: "Blue",
  O: "Orange",
  Y: "Yellow",
  G: "Green",
  V: "Violet",
  R: "Red",
  M: "Grey"
}

const lineStyle = {
  display: "flex",
  flexDirection: "row"
}

const render = (line, i) => {
  return (
    <div key={i} style={lineStyle}>
      { line.map((v, k) => ( <div key={k} style={box(v)}/> )) }
    </div>
  )
}

const box = (c) => {
  return {
    width: "10px",
    height: "10px",
    backgroundColor: color[c]
  }
}

const Board = ({display}) => (
  <div>
    {display.map((value, key) => render(value, key))}
  </div>

)
export const Card = ({ player }) => (
  <div style={style}>
    <h6>{player.name}</h6>
    <Board display={player.display}/>
  </div>
)


// const mapStateToProps = (state, ownProps) => ({
//   player: ownProps.player
// })

// //export default Card
// export default connect(mapStateToProps)(Card)
