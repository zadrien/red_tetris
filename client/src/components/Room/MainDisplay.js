import React from 'react'
import { connect } from 'react-redux'

import GameOver from './Overlay/GameOver'
import Host from './Overlay/Host'

const color = {
	".": "Black",
	C: "Cyan",
	B: "Blue",
	O: "Orange",
	Y: "Yellow",
	G: "Green",
	V: "Violet",
	R: "Red",
	M: "Grey"
}

const box = (c) => ({backgroundColor: color[c]})

const render = (line, index) => (
	<div key={index} className="board-row">
		{ line.map((v, k) => (
			<div key={k} className="board-box" style={box(v)}/>
		))}
	</div>
)

const Board = ({ display, Overlay }) => (
	<div className="board">
		{Overlay ? <Overlay/> : null}
		{display.map((v, k) => render(v, k))}
	</div> 
)

export const Display = ({room, display, isStarted}) => {
	if (!display)
		return <div className="board"/>
	if (room.hasOwnProperty('winner'))
		return (<Board display={display} Overlay={() => <GameOver/>}/>)
	if (!isStarted)
		return <Board display={display} Overlay={() => <Host/>}/>
	return <Board display={display} />
}


const mapStateToProps = (state) => ({
	room: state.room,
	display: state.room.display,
	isStarted: state.room.start,
	
})

export default connect(mapStateToProps)(Display)
