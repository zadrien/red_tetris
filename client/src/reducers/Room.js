const room = (state = {}, action = {}) => {
	var room = Object.assign({}, state.room)
	
	switch (action.state) {
	case "JOINED":
		if (!action.result.room)
			return state
		return Object.assign({}, state, { menu: "ROOM", room: action.result.room })

	case "START":
		if (!action.result)
			return state
		room.isOpen = action.result.start
		return Object.assign({}, state, { room: room/*ADD SOMETHING*/})
		
	case "DISPLAY":
		room.display = action.result
		return Object.assign({}, state, { room: room/*ADD SOMETHING*/})
		
	case "HOST":
		room.host = action.result.host
		return Object.assign({}, state, { room: room/*ADD SOMETHING*/})
		
	case "PLAYERS":
		var obj = action.result.filter( (e) => {
			return e.name !== state.user.name
		})
		if (!obj)
			return state
		room.players = obj
		return Object.assign({}, state, { room: room/*ADD SOMETHING*/})
		
	case "GAMEOVER":
		room.winner = (action.result.winner)
		return Object.assign({}, state, { room: room/*ADD SOMETHING*/})
		
	case "RESET":
		delete room.winner
		return Object.assign({}, state, { room: room })
		
	case "QUIT":
		return Object.assign({}, state, { room: undefined, menu: "LISTING" })
		
	default:
		return state
	}
}


export default room
