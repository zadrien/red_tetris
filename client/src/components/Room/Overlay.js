import React, { useEffect } from 'react'
import { removeOverlay } from '../../actions/Room'

import './style.css'
import { connect } from 'react-redux'
const displayStyle = {
	// zIndex: 9,
	// position: "absolute",
	marginTop: ".5px",
	width: "360px",
	height: "720px",
	//backgroundColor: "grey",
	textAlign: "center",
	padding: "5px",
	// opacity: 0.8
}

const OverlayDisplay = ({ room, onClick}) => {
	useEffect(() => {
		let interval = setTimeout(() => onClick(), 3000)

		return () => {
			clearInterval(interval)
		}
	}, [])

	return (
		<div className="board-overlay transparent d-flex col space-around" style={displayStyle}>
		<h6 className="title big">{ room.winner === true ? "You won" : "You lost"}</h6>
		<button className="bob-btn" onClick={onClick}>OK</button>
		</div>
	)
}
export const Overlay = ({ room, onClick }) => {
	if (!room.hasOwnProperty("winner"))
		return null
	return <OverlayDisplay room={room} onClick={onClick}/>
}

const mapStateToProps = (state) => {
	if (state.room)
		return {
			room: state.room
		}
	return {
		room: undefined,
	}
}

const mapDispatchToProps = (dispatch) => ({
	onClick: () => {
		dispatch(removeOverlay())
	}
})

export default connect(mapStateToProps, mapDispatchToProps)(Overlay)