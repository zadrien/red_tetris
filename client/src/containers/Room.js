import React, {useEffect} from 'react';
import { connect } from 'react-redux';
import { emitQuit, emitMove } from '../actions/Room'

import Board from '../components/Room/MainDisplay'
import Players from '../components/Room/Player'
import Host from '../components/Room/host'

import '../global.css';

export const Room = ({ room, onStartUp, enableController, disableController, onLeave }) => {
	useEffect(() => {
		onStartUp(room)
		enableController()

		return () => {
			disableController()
		}
	}, [])

	if (room)
		return (
			<div>
				<div className="align-center width-100">
				<h1 className="title medium">{room.name}</h1>
			</div>
			<div className="d-flex just-center mx-auto width-25">
				<Host />
				<div className="bob-btn secondary" onClick={() => onLeave(room)}>Leave</div>
			</div>
			<div className="p-2 width-100 d-flex row just-center mx-auto">
				<Board/>
				<Players />
			</div>
	
			</div>
		)
	return null
}

const mapStateToProps = (state) => ({
		room: state.room
})

const mapDispatchToProps = (dispatch) => () => {
	const controller = (e) => {
		if ([32, 37, 38, 39, 40].indexOf(e.keyCode) > -1)
				e.preventDefault()
		var key = e.keyCode
		if (key === 38) { // UP
			console.log("UP")
			dispatch(emitMove("UP"));
		} else if (key === 39) { // RIGHT
			dispatch(emitMove("RIGHT"));
		} else if (key === 40) { // DOWN
			dispatch(emitMove("DOWN"));
		} else if (key === 37) { // LEFT
			dispatch(emitMove("LEFT"));
		} else if (key === 32) { // SPACE
			dispatch(emitMove("SPACE"));
		}
	}

	const changeURL = (room) => {
		window.history.pushState(null,null, `#${room.id}`)
		document.title = room.name.toUpperCase()
	}


	return ({
		onStartUp: (e) => {
			changeURL(e)
		},
		enableController: () => {
			window.addEventListener('keydown', controller)
		},
		disableController: () => {
			window.removeEventListener('keydown', controller)
		},
		onLeave: (room) => {
			dispatch(emitQuit(room))
			window.history.back()
			document.title = "Red Tetris"
		}
	})
}
export default connect(mapStateToProps, mapDispatchToProps)(Room);
