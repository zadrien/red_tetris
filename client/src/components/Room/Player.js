import React from 'react'
import { connect } from 'react-redux';
import { Card } from './Card'

const style = {
	display: "flex",
	flexWrap: "wrap",
	marginLeft: "5px",
}

export const Players = ({ players }) => {
	if (!players)
		return null
	return (
		<div className="d-flex row" style={style}>
			{players.map((p, index) => (
				<Card key={index} player={p}/>
			))}
		</div>
	)
}

const mapStateToProps = (state) => ({
	players: state.room.players
})

export default connect(mapStateToProps)(Players)
