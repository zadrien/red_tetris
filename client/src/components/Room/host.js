import React from 'react'
import { connect } from 'react-redux'
import { startGame, emitStart, onDisplay, onPlayers } from '../../actions/Room'
import { Overlay } from './Overlay'

const displayStyle = {
	marginTop: ".5px",
	width: "360px",
	height: "720px",
	textAlign: "center",
	padding: "5px",
}

const WaitingHost = () => (
	<div>
		<h6 className="title big">Pending...</h6>
	</div>
)

const HostOverlay = ({onClick, id}) => (
	<div>
		<h6 className="title big">Ready ?</h6>
		<div className={`bob-btn secondary`} onClick={() => onClick(id)}>Start</div>
	</div>
)

export const Host = ({host,  onClick, id}) => {
	if (!host)
		return <Overlay elem={() => <WaitingHost/>}/>
	return <Overlay elem={() => <HostOverlay onClick={onClick} id={id}/>}/>
}

const mapStateToProps = (state, ownProps) => ({
	host: state.room.host,
	start: state.room.start,
	id: state.room.id
})

const mapDispatchToProps = (dispatch, ownProps) => ({ 
	onClick: (id) => {
		dispatch(emitStart(id))
		dispatch(onDisplay())
		dispatch(onPlayers())
		dispatch(startGame(true))
	}
})

export default connect(mapStateToProps, mapDispatchToProps)(Host)
