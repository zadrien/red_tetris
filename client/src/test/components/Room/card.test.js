import React from 'react'
import { expect } from 'chai'
import { shallow } from 'enzyme'

import { Card } from '../../../components/Room/Card'

function setup() {
  const props = {
    player: {
      name: "testName",
      display: [
        [ '.', '.', '.', '.', '.', '.', '.', '.', '.', '.' ],
        [ '.', '.', '.', '.', '.', '.', '.', '.', '.', '.' ],
        [ '.', '.', '.', '.', '.', '.', '.', '.', '.', '.' ],
        [ '.', '.', '.', '.', '.', '.', '.', '.', '.', '.' ],
        [ '.', '.', '.', '.', '.', '.', '.', '.', '.', '.' ]
      ]
    }
  }

  const enzymeWrapper = shallow(<Card {...props}/>)
  return {
    props,
    enzymeWrapper
  }
}

describe('Card BDD', () => {
  it('should render Card component', function() {
    const { enzymeWrapper } = setup()

    expect(enzymeWrapper.find('h6').text()).to.equal('testName')
    const board = enzymeWrapper.find('Board').props()
    expect(board.display).to.be.a('Array')
  })
})
