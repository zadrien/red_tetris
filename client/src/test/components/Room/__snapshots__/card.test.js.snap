// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card TDD should render  1`] = `
<div
  className="mini-board"
>
  <h6>
    testName
  </h6>
  <div>
    <div
      className="board-row"
    >
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
    </div>
    <div
      className="board-row"
    >
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
    </div>
    <div
      className="board-row"
    >
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
    </div>
    <div
      className="board-row"
    >
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
    </div>
    <div
      className="board-row"
    >
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
      <div
        className="mini-board-box"
        style={
          Object {
            "backgroundColor": "Black",
          }
        }
      />
    </div>
  </div>
</div>
`;
